lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -17,17 +17,94 @@
+    "enable-partial-specialization", llvm::cl::init(false),
+    llvm::cl::desc("Enable partial specialization of generic functions"));
+
+
+/// If set, then generic specialization tries to specialize using
+/// all substitutions, even if their replacement types are generic.
+/// Otherwise specialize only those generic type parameters which are
+/// substituted by concrete types.
+static bool CanSpecializeConcreteAndGenericSubstitutions = true;
+
+/// If set, generic substitutions are substituted into function signatures.
+/// Given a substitution, U = G<V>, function signature 'func foo<U>(x: U)'
+/// becomes 'func foo<V>(x: G<V>)'.
+///
+/// Only has impact if SpecializeConcreteAndGenericSubstitutions is set.
+/// FIXME: This is a temporary flag that will be removed once the implementation
+/// of the partial specialization is ready.
+static bool SupportGenericSubstitutions = false;
+
+/// Apply certain heuristics to decide of a given generic substitution
+/// should be used for the generic specialization. As an optimization, the
+/// compiler may decide that a given generic substitution is not profitable
+/// and can decide to avoid substituting it into a function signature.
+///
+/// Only has impact if SpecializeConcreteAndGenericSubstitutions and
+/// SupportGenericSubstitutions are set.
+/// FIXME: This is a temporary flag that will be removed once the implementation
+/// of the partial specialization is ready.
+static bool OptimizeGenericSubstitutions = false;
+
+/// If set to false, fully concrete substitutions use the same code path
+/// as partial specializations, which eliminates a need to special case
+/// the full specialization.
+/// Currently it is disabled, because ArchetypeBuidler cannot properly
+/// handle same-type requirements to concrete types in certain cases, which
+/// occur in the stdlib.
+/// rdar://29333056
+/// FIXME: This is a temporary flag that will be removed once the implementation
+/// of the partial specialization is ready.
+static bool ShortcutFullSpecialization = true;
+
 // Max depth of a bound generic which can be processed by the generic
 // specializer.
 // E.g. the depth of Array<Array<Array<T>>> is 3.
 // No specializations will be produced, if any of generic parameters contains
 // a bound generic type with the depth higher than this threshold 
 static const unsigned BoundGenericDepthThreshold = 50;
 
+class QueryConformance {
+  ModuleDecl *const Mod;
+public:
+  QueryConformance(ModuleDecl* Mod): Mod(Mod) {}
+
+  Optional<ProtocolConformanceRef>
+  operator()(CanType original, Type replacement, ProtocolType *protoType) {
+    auto proto = protoType->getDecl();
+    auto C = Mod->lookupConformance(replacement, proto, nullptr);
+    if (C.hasValue())
+      return C.getValue();
+
+    if (!replacement->hasTypeParameter() && !replacement->hasArchetype()) {
+      // It is a concrete type.
+      auto nominal = replacement->getNominalOrBoundGenericNominal();
+      if (nominal) {
```

Comment:
This looks like it's duplicating your new utility function in ProtocolConformanceRef